crates/sequencer/src/sequencer.rs

```diff
@@ -1,17 +1,23 @@
-use crate::config::SequencerConfig;
+use blockifier::block_context::BlockContext;
 use blockifier::state::state_api::{State, StateReader};
 
 /// Sequencer is the main struct of the sequencer crate.
 /// Using a trait bound for the state allows for better
 /// speed, as the type of the state is known at compile time.
-pub struct Sequencer<S: State + StateReader> {
-    pub config: SequencerConfig,
+pub struct Sequencer<S>
+where
+    for<'a> &'a mut S: State + StateReader,
+{
+    pub context: BlockContext,
     pub state: S,
 }
 
-impl<S: State + StateReader> Sequencer<S> {
+impl<S> Sequencer<S>
+where
```

Comment:
We should add more comments on complex rust bounds and generics, i.e.

`For all lifetimes 'a, a mutable reference to S (&'a mut S) must implement both the State and StateReader traits.`

https://doc.rust-lang.org/nomicon/hrtb.html

btw why do we need such a deep magic?

Reply:
So basically I have to implement State and StateReader not for S but for &mut S, this means that I need to bring a lifetime in, and apparently I need to use for<'a>, I am not a expert on this but I guess it just means that I don't impose restrictions on the lifetime, just saying that for any lifetime, the reference to State must implement State and StateReader (as you said).